### PR TITLE
jobs: introduce system.job_info table and helpers

### PR DESCRIPTION
**Pull request by dt**
See commits.

Design: https://github.com/cockroachdb/cockroach/pull/82638

Epic: None.

### CODE REVIEW

**Review by postamar — COMMENTED** [review]
I feel like there must be some issue or RFC or other documentation providing context as to what purpose this table will serve, could you provide a link please? I'm not sure where to look, I've looked in the RFCs folder and the Jobs board to no avail.

> Reviewed 10 of 10 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 41 of 41 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-:-N7pbtI33143ZlvPjxZt:bwqa4yh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @samiskin, and @stevendanna)

___
*[`pkg/jobs/registry.go` line 975 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pXhbjE6KypJCMKeO5:-N7pXhbjE6KypJCMKeO6:b-istjvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6e407cf1c8f78a06a4510042e0f17b24ab98e15/pkg/jobs/registry.go#L975)):*
> ```Go
> 		}
> 		if nDeletedInfos, err = r.ex.Exec(
> 			ctx, "gc-jobs", nil /* txn */, infoStmt, toDelete,
> ```

Shouldn't this be in the same transaction as above? We don't have an explicit foreign key but I'm slightly worried about logical referential integrity, if that makes sense.
___
*[`pkg/jobs/registry.go` line 977 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pY_YHAjbGtxjM1kTH:-N7pY_YHAjbGtxjM1kTI:bae44qj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6e407cf1c8f78a06a4510042e0f17b24ab98e15/pkg/jobs/registry.go#L977)):*
> ```Go
> 			ctx, "gc-jobs", nil /* txn */, infoStmt, toDelete,
> 		); err != nil {
> 			return false, 0, errors.Wrap(err, "deleting old jobs")
> ```

nit: is there any value in giving this query and error message wrapper distinct string literals, versus the `system.jobs` query?
___
*[`pkg/jobs/registry.go` line 682 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pZP_H7KPQnKRhLmBH:-N7pZP_H7KPQnKRhLmBI:b-xsrilv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9d58389b980f4a15c7f33bc0cf3444e48252a73/pkg/jobs/registry.go#L682)):*
> ```Go
> // IterateJobInfo iterates though the info records for a given job and info key
> // prefix.
> // TODO(dt): use pagination in query.
> ```

Shouldn't this be tracked in an issue which is marked as a GA-blocker? It seems important enough. Please ignore if this is the case already.
___
*[`pkg/jobs/registry.go` line 744 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7p_GTW8E1t45X2rkGv:-N7p_GTW8E1t45X2rkGw:b6kg8hn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9d58389b980f4a15c7f33bc0cf3444e48252a73/pkg/jobs/registry.go#L744)):*
> ```Go
> 		ctx, "job-info-write", txn,
> 		sessiondata.InternalExecutorOverride{User: username.NodeUserName()},
> 		"DELETE FROM system.job_info WHERE job_id = $1 AND info_key = $2",
> ```

Why is `written` in the primary key at all, if there's effectively a 1:1 relationship between it and the `(job_id, info_key)` pair? This deserves commentary, otherwise one might wonder: why not simply rely on the `crdb_internal_mvcc_timestamp` column instead?
___
*[`pkg/jobs/registry.go` line 755 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pZjm51snp6TuoAKAn:-N7pZjm51snp6TuoAKAo:b-9vly7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9d58389b980f4a15c7f33bc0cf3444e48252a73/pkg/jobs/registry.go#L755)):*
> ```Go
> 		ctx, "job-info-write", txn,
> 		sessiondata.InternalExecutorOverride{User: username.NodeUserName()},
> 		`INSERT INTO system.job_info (job_id, info_key, written, value) VALUES ($1, $2, now(), $3)`,
> ```

nit: is the inclusion of `written` deliberate here? I feel it's fine to remove and rely on its default value, it's a common-enough pattern.
___
*[`pkg/sql/catalog/systemschema/system.go` line 682 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pWGnf08paosA4PDBQ:-N7pWGnf08paosA4PDBR:b-8a4htw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8557a09291a34eab0d096954cb2168f12b52ce3e/pkg/sql/catalog/systemschema/system.go#L682)):*
> ```Go
> CREATE TABLE system.job_info (
> 	job_id INT8 NOT NULL,
> 	info_key BYTES NOT NULL,
> ```

I'm missing context here on what this is. Can you please explain, either in the code or in an issue? Am I correct in guessing that this table is a deliberately unstructured key-value store for each job?
___
*[`pkg/upgrade/upgrades/system_job_info_test.go` line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pXOLg1aRucEX8iVvP:-N7pXOLg1aRucEX8iVvQ:bt86h1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/8557a09291a34eab0d096954cb2168f12b52ce3e/pkg/upgrade/upgrades/system_job_info_test.go#L28)):*
> ```Go
> )
> 
> func TestSystemJobInfoMigration(t *testing.T) {
> ```

nit: FYI we call them "upgrades" now, "migration" is so 2021. 
___
*[`pkg/upgrade/upgrades/system_job_info_test.go` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pXDlq9SOQVzzQHLQ3:-N7pXDlq9SOQVzzQHLQ4:b78rns6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8557a09291a34eab0d096954cb2168f12b52ce3e/pkg/upgrade/upgrades/system_job_info_test.go#L41)):*
> ```Go
> 	tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{
> 		ServerArgs: base.TestServerArgs{
> 			Settings: settings,
> ```

Is this necessary at all? I'm not seeing this pattern in other upgrade tests.
___
*[`pkg/upgrade/upgrades/system_job_info_test.go` line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pXC6L1WwtXD0oXSe9:-N7pXC6L1WwtXD0oXSeA:b-589zjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8557a09291a34eab0d096954cb2168f12b52ce3e/pkg/upgrade/upgrades/system_job_info_test.go#L45)):*
> ```Go
> 				Server: &server.TestingKnobs{
> 					DisableAutomaticVersionUpgrade: make(chan struct{}),
> 					BinaryVersionOverride:          clusterversion.ByKey(clusterversion.SystemPrivilegesTable - 1),
> ```

Shouldn't the key here be `CreateSystemJobInfoTable` instead of `SystemPrivilegesTable`? See also elsewhere in this test definition. I'm guessing that's why the test is failing presently.

File: pkg/jobs/registry_test.go
Comment (Xiang-Gu) [inline pkg/jobs/registry_test.go:1155]: remove this line?

File: pkg/jobs/registry_test.go
Comment (Xiang-Gu) [inline pkg/jobs/registry_test.go:1172]: nit: "is now found with "

File: pkg/sql/catalog/systemschema/system.go
Comment (Xiang-Gu) [inline pkg/sql/catalog/systemschema/system.go:685]: why is `written` part of the primary key, given that the accessor helpers (in commit 3) look up by (job_id, info_key)?

I guess I'll need to read the implementation to see how we handle the case when there are >1 rows with the same (job_id, info_key) but different 'written' time.

**Comment by ajwerner** [comment]
@postamar https://github.com/cockroachdb/cockroach/pull/82638

**Comment by postamar** [comment]
Thanks for this @ajwerner. This helps clear up a lot of the questions I had regarding the `written` and `info_key` columns, please interpret my review comments accordlngly.

**Comment by postamar** [comment]
Frankly I'd prefer we didn't pass naked `[]byte`s to and from the accessors like `GetJobInfo`  and instead had strongly-typed accessors for each job (sub) type, ideally relying on some specific protobuf message definitions. These would, for the time being, be placeholder empty messages. You probably want protobufs anyway, to guarantee encoding stability across cluster versions and because it's easy to design these messages in a way that allows straightforward definition of prefixes.

**Review by dt — COMMENTED** [review]
I very much don't want a central library handing out ImportInfo / RestoreInfo / SchemaChangeInfo / etc; we shouldn't need to edit the core jobs code dependency root every time a leaf job changes a field in its persisted data (we do today and it's not good). I _want_ a "I want to store this thing and I'll ask for it later" API where the library is agnostic to what we're storing.

That said, I could see switching from []byte to proto.Message or a `message job.InfoRecord { types.Any payload = 1}` wrapper proto or something (though I've started to worry about types.Any and refactoring changing persisted type names more lately).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-:-NE35W95BU4l72E7qYw4:b-11g5kg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @postamar, @samiskin, and @stevendanna)

___
*[`pkg/jobs/registry.go` line 975 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pXhbjE6KypJCMKeO5:-NE37Af9-pW-x6gQr6gs:b-jx9833) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6e407cf1c8f78a06a4510042e0f17b24ab98e15/pkg/jobs/registry.go#L975)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't this be in the same transaction as above? We don't have an explicit foreign key but I'm slightly worried about logical referential integrity, if that makes sense.
</blockquote></details>

Great question!

I wanted an FK but ultimately we can't have one, since we want to allow job creators to stage info rows *before* creating a job row, and indeed even to do so in separate transactions, before the txn that creates the job row. Allowing this -- that info rows can be created even without a job row -- is pretty important since we want to do as little as possible in that job-creating transaction since that causes job. So we've accepted that there may be at least temporary inconsistency between these, and indeed more than that -- a job creator could crash after some of its staging, with some info rows and then fail to create a job and not clean them up. So we might need a recon cleanup loop of last resort anyway (though not urgently -- a couple orphaned info rows are pretty harmless). Given that, I'm inclined not to pile them into the same txn here., as long as we delete from the jobs first, then the infos (which we do), just to minimize the risk of something in the info deletion blocking job deletion. 
___
*[`pkg/jobs/registry.go` line 977 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pY_YHAjbGtxjM1kTH:-NE3C0NWDouuQ2MRAZWu:b-r4rvcc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6e407cf1c8f78a06a4510042e0f17b24ab98e15/pkg/jobs/registry.go#L977)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: is there any value in giving this query and error message wrapper distinct string literals, versus the `system.jobs` query?
</blockquote></details>

I think our errors library usually gets us the stack anyway so I don't think so?
___
*[`pkg/jobs/registry.go` line 682 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pZP_H7KPQnKRhLmBH:-NE36E6b37253r_x9cQT:b-muyqnt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9d58389b980f4a15c7f33bc0cf3444e48252a73/pkg/jobs/registry.go#L682)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't this be tracked in an issue which is marked as a GA-blocker? It seems important enough. Please ignore if this is the case already.
</blockquote></details>

Only an issue if/once it has a production caller who is iterating a non-trivial number of rows, so I'd say it isn't a blocker until then -- i.e. if initial callers just persist 2 info rows for each job (payload and progress) then this can wait.
___
*[`pkg/jobs/registry.go` line 744 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7p_GTW8E1t45X2rkGv:-NE36aFQEmYF2RB5I2Pd:b5c8azs) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9d58389b980f4a15c7f33bc0cf3444e48252a73/pkg/jobs/registry.go#L744)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Why is `written` in the primary key at all, if there's effectively a 1:1 relationship between it and the `(job_id, info_key)` pair? This deserves commentary, otherwise one might wonder: why not simply rely on the `crdb_internal_mvcc_timestamp` column instead?
</blockquote></details>

As you found/mentioned above, this is mentioned in the design doc but we're doing our own mvcc since we know these job info (logical) rows get lots of "in-place" updates in long-running jobs so we're splitting that into separate crdb table (physical) rows, since KV and SQL don't allow splitting between kv mvcc revisions of a single sql row, but we want to be able to split between revisions of a job row.
___
*[`pkg/jobs/registry.go` line 755 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pZjm51snp6TuoAKAn:-NE3719g4q2Xiw8SCGeg:bvhqln4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9d58389b980f4a15c7f33bc0cf3444e48252a73/pkg/jobs/registry.go#L755)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: is the inclusion of `written` deliberate here? I feel it's fine to remove and rely on its default value, it's a common-enough pattern.
</blockquote></details>

Personal preference: I like to see what I'm writing and not need to pull up the schema to find the default defs.

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84866)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @baoalvin1, @postamar, @samiskin, @stevendanna, and @Xiang-Gu)

___
*[`pkg/sql/catalog/systemschema/system.go` line 685 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7X5P0jz7IBBZQcqolw-r1-685:-NFnuwqb9M4SE8_fCaPY:b3s6z5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/8557a09291a34eab0d096954cb2168f12b52ce3e/pkg/sql/catalog/systemschema/system.go#L685)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

why is `written` part of the primary key, given that the accessor helpers (in commit 3) look up by (job_id, info_key)?

I guess I'll need to read the implementation to see how we handle the case when there are >1 rows with the same (job_id, info_key) but different 'written' time. 
</blockquote></details>

I added a comment inline here summarizing the reasoning explored in the RFC.

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84866)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @baoalvin1, @dt, @postamar, @samiskin, @stevendanna, and @Xiang-Gu)

___
*[`pkg/sql/catalog/systemschema/system.go` line 682 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pWGnf08paosA4PDBQ:-NGRo5_U5iPZlbk7E5JU:b-w12guy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8557a09291a34eab0d096954cb2168f12b52ce3e/pkg/sql/catalog/systemschema/system.go#L682)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'm missing context here on what this is. Can you please explain, either in the code or in an issue? Am I correct in guessing that this table is a deliberately unstructured key-value store for each job?
</blockquote></details>

The RFC linked in the PR now is pretty explicit about the fact being arbitrary keys is deliberate. 
___
*[`pkg/jobs/registry_test.go` line 1155 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7X5P0k6tlasTbnJsC_-r3-1155:-NGRo_LKACmfdamDRKrd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9d58389b980f4a15c7f33bc0cf3444e48252a73/pkg/jobs/registry_test.go#L1155)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

remove this line?
</blockquote></details>

Done.
___
*[`pkg/jobs/registry_test.go` line 1172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7X5P0k6tlasTbnJsC_-r3-1172:-NGRo_hl8AoBzpA9jGoc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9d58389b980f4a15c7f33bc0cf3444e48252a73/pkg/jobs/registry_test.go#L1172)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

nit: "is now found with "
</blockquote></details>

Done.
___
*[`pkg/upgrade/upgrades/system_job_info_test.go` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pXDlq9SOQVzzQHLQ3:-NGRnTCC03I_GlsNsLsO:bp2re6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/8557a09291a34eab0d096954cb2168f12b52ce3e/pkg/upgrade/upgrades/system_job_info_test.go#L41)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is this necessary at all? I'm not seeing this pattern in other upgrade tests.
</blockquote></details>

I think all the upgrade tests do this, so they can initialize the version to before they run.
___
*[`pkg/upgrade/upgrades/system_job_info_test.go` line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84866#-N7pXC6L1WwtXD0oXSe9:-NGRnTuMB0kZjkTZ4v4a:bmfrqvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8557a09291a34eab0d096954cb2168f12b52ce3e/pkg/upgrade/upgrades/system_job_info_test.go#L45)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't the key here be `CreateSystemJobInfoTable` instead of `SystemPrivilegesTable`? See also elsewhere in this test definition. I'm guessing that's why the test is failing presently.
</blockquote></details>

The start should be before the migration so that the test can assert that running the migration actually advances from before to after states as expected; at the time I wrote this SystemPrivilegesTable was the most recent version so that's what it used as its before, but now I makes more sense to be V22_2.

File: pkg/jobs/registry.go
Comment (postamar) [inline pkg/jobs/registry.go:1095]: uber nit: `s/gc-jobs/gc-job-info/` ?

File: pkg/jobs/registry.go
Comment (postamar) [inline pkg/jobs/registry.go:1098]: uber nit: likewise, `s/old jobs/old job infos/`?

**Review by postamar — COMMENTED** [review]
I understand that you're still fighting CI test failures, but this seems close enough.

> I very much don't want a central library handing out ImportInfo ... [...] That said, I could see switching from []byte to proto.Message

Fair enough. Also, the []byte -> proto.Message doesn't need to happen now, if ever.

File: pkg/upgrade/upgrades/system_job_info_test.go
Comment (dt) [reply pkg/upgrade/upgrades/system_job_info_test.go:45]: Eh, I viscerally dislike the vFoo-1 "trick": it has a the terrible effect of meaning that you can delete the preceding  versions, such that vFoo is the 0th element, and the compiler is perfectly happy, since you deleted those previous things and their usages just as you should, but then this -1 at runtime is an out-of-bound crash, since it was _implicitly_ referring to a thing you deleted, but in a way that was hidden from the compiler. I hate it, and I want to lead by example here and show that you should use actual symbol, by name, of the before version.

File: pkg/sql/pgwire/pgwire_test.go
Comment (dt) [reply pkg/sql/pgwire/pgwire_test.go:573]: whoops, yeah, bad rebase (didn't even compile) from when this started on 22.2 when we didn't have that const.

File: pkg/sql/logictest/testdata/logic_test/distsql_enum
Comment (dt) [reply pkg/sql/logictest/testdata/logic_test/distsql_enum:75]: Whoops, yeah, looks like the --rewrite needed fix all the overly specific ranges stuff that changes when you add a system table ran into the non-deterministic plan stuff. Nice catch.

File: pkg/upgrade/upgrades/system_job_info_test.go
Comment (dt) [reply pkg/upgrade/upgrades/system_job_info_test.go]: Nope, you're right; it wasn't working. I have no idea if/how it was working back when I wrote this in June, but it certainly wasn't as rebased today. Fixed.

File: pkg/sql/tests/system_table_test.go
Comment (dt) [reply pkg/sql/tests/system_table_test.go]: Ah, nice!

**Comment by dt** [comment]
> still fighting CI test failures

Think that's the last of them, for this rebase at least!

File: pkg/upgrade/upgrades/system_job_info_test.go
Comment (postamar) [reply pkg/upgrade/upgrades/system_job_info_test.go:45]: Ok, can you file an issue to track the desire to adopt that pattern throughout all upgrades, please? Seems worth it.

File: pkg/upgrade/upgrades/system_job_info_test.go
Comment (postamar) [reply pkg/upgrade/upgrades/system_job_info_test.go]: This is probably not deliberate on your part but this still doesn't seem fixed to me. The assertion checking that the table is not present before the upgrade runs has been commented out. I'm also not seeing any code to "un-bootstrap" the jobs_info table, which you need to simulate the v22.2 bootstrapped system schema.

Comment (dt) [reply pkg/upgrade/upgrades/system_job_info_test.go]: Oh, no, you're not missing anything; there's nothing that un-bootstraps it to test that the migration _actually_ creates a table, but as far as I can tell, this is in line with all the tests of `createSystemTable`-based upgrades, as there is no way to bootstrap the system schema to the prior state, just the version. So I suppose these tests are all just proving the upgrade is idempotent w.r.t the schema change, and this one is too. Added a comment to this effect.

File: pkg/upgrade/upgrades/system_job_info_test.go
Comment (dt) [reply pkg/upgrade/upgrades/system_job_info_test.go:45]: https://github.com/cockroachdb/cockroach/issues/91675

File: pkg/upgrade/upgrades/system_job_info_test.go
Comment (ajwerner) [reply pkg/upgrade/upgrades/system_job_info_test.go]: That seems right. There was talk once upon a time to version gate bootstrap to control for this. I think @Xiang-Gu added a roachtest that does this check in https://github.com/cockroachdb/cockroach/blob/ee061ddfcbe5f45f0e210927d50dcd4c8a55baf4/pkg/cmd/roachtest/tests/validate_system_schema_after_version_upgrade.go#L25 which does an upgrade and records `SHOW CREATE ALL TABLES` for the `system` database and then does that on a cluster bootstrapped at the current version and makes sure they are the same.

Comment (postamar) [reply pkg/upgrade/upgrades/system_job_info_test.go]: Oh well, TIL!

Comment (dt) [reply pkg/upgrade/upgrades/system_job_info_test.go]: yeah, that's a good test and I'm glad we have it. This test isn't, but it is in-line with the others, so I think it is as-expected.

**Comment by dt** [comment]
Bump; this is ready for another (last? maybe?) round of reviews

File: pkg/jobs/registry.go
Comment (ajwerner) [inline pkg/jobs/registry.go:737]: This one should get some sort of batching API.

File: pkg/upgrade/upgrades/precondition_before_starting_an_upgrade_external_test.go
Comment (ajwerner) [inline pkg/upgrade/upgrades/precondition_before_starting_an_upgrade_external_test.go:122]: I think this test evolved such that this line is still legit, no?

File: pkg/jobs/registry.go
Comment (ajwerner) [inline pkg/jobs/registry.go:686]: What do you mean by this? The query is streamed. Do you mean, allow the user to resume if they want to paginate?

File: pkg/jobs/registry.go
Comment (ajwerner) [inline pkg/jobs/registry.go:658]: One thing that's sort of nice about the job writing APIs are that they observe the session claim if there is one (which there is on the job object during execution) and will fail if the job has been preempted. We'll lose this in this PR. I also am rapidly falling out of love with these methods on the registry which take a txn. Can I nudge you towards making a new structure which holds onto a reference to the job and the txn and then has methods to read and write these keys?

maybe like:

```go

func (j *Job) InfoStorage(txn *kv.Txn) InfoStorage

type InfoStorage struct {
    j *Job
    txn *kv.Txn
}

func (is InfoStore) Get(key []byte) (value []byte, exists bool, _ error) 
```

**Review by msbutler — APPROVED** [review]
Hope no one slips another system table in before you merge :D

**Review by ajwerner — APPROVED** [review]
I'm cool with the wait-and-see approach

**Review by stevendanna — APPROVED** [review]
Thanks for working on this.

It will be nice to start playing with this.

I think we will want to think about the semantics of updating a job_info record with or without the lease, but I'm happy to leave that to a follow up if it ends up being needed, as you already indicated.

File: pkg/jobs/registry.go
Comment (dt) [reply pkg/jobs/registry.go:706]: the primary index (only) is the one I want it to hit, but I wanted that span to be tighter -- in theory the optimizer should be able to know to constraint to /98/foo, not just /98

**Pull request merged**
